panther_analysis_tool/detection.py

```diff
@@ -189,3 +249,597 @@ def get_module(  # pylint: disable=arguments-differ
+        self.detection_type = config.get("analysisType", default_type)
+        self.detection_id = config["id"]
+        self.detection_version = config["versionId"]
+
+        # TODO: severity and other detection metadata is not passed through when we run tests.
+        #       https://app.asana.com/0/1200360676535738/1200403272293475
+        if "severity" in config:
+            self.detection_severity = config["severity"]
+        else:
+            self.detection_severity = None
+
+        if not ("dedupPeriodMinutes" in config) or not isinstance(
+            config["dedupPeriodMinutes"], int
+        ):
+            self.detection_dedup_period_mins = DEFAULT_RULE_DEDUP_PERIOD_MINS
+        else:
+            self.detection_dedup_period_mins = config["dedupPeriodMinutes"]
+
+        if not ("tags" in config) or not isinstance(config["tags"], list):
+            self.detection_tags: List[str] = list()
+        else:
+            config["tags"].sort()
+            self.detection_tags = config["tags"]
+
+        if "reports" not in config:
+            self.detection_reports: Dict[str, List[str]] = dict()
+        else:
+            # Reports are Dict[str, List[str]]
+            # Sorting the List before setting it
+            for values in config["reports"].values():
+                values.sort()
+            self.detection_reports = config["reports"]
+
+        self._setup_exception = None
+        try:
+            self._module = self._load_detection(self.detection_id, config)
+            if not hasattr(self._module, self.matcher_function):
+                raise AssertionError(
+                    f"detection needs to have a method named '{self.matcher_function}'"
+                )
+        except Exception as err:  # pylint: disable=broad-except
+            self._setup_exception = err
+            return
+
+        self._default_dedup_string = "defaultDedupString:{}".format(self.detection_id)
+        self._auxiliary_function_definitions = self._check_defined_functions()
+
+    @staticmethod
+    def _load_detection(identifier: str, config: Mapping) -> ModuleType:
+        """Load the detection code as Python module.
+        Code can be provided as raw string or a path in the local filesystem.
+        """
+        has_raw_code = bool(config.get("body"))
+
+        importer: Optional[BaseImporter] = None
+        if has_raw_code:
+            resource = config["body"]
+            importer = RawStringImporter(tmp_dir=_DETECTION_FOLDER)
+        else:
+            resource = config["path"]
+            importer = FilesystemImporter()
+
+        return importer.get_module(identifier, resource)
+
+    def _check_defined_functions(self) -> Dict[str, bool]:
+        function_definitions = {}
+        for name in AUXILIARY_FUNCTIONS:
+            function_definitions[name] = self._is_function_defined(name)
+        return function_definitions
+
+    @property
+    def matcher_function(self) -> str:
```

Comment:
Maybe `matcher_function_name` ? Also, you can make `Detection` an Abstract Base class (if you decide to not rename `Rule -> Detection` and proceed with inheritance in this PR) so this would become an [abstract property](https://docs.python.org/3/library/abc.html#abc.abstractproperty). Sub-classes must define these methods (the IDE would even warn you!), otherwise they cannot be instantiated, plus the abstract class cannot be instantiated too, so there's no confusion there.